### DISCUSSION TITLE
FeatureExtraction Example

### DISCUSSION THREAD

**Discussion by ahabdallah27** [NeurodataWithoutBorders/helpdesk#74] [Q&A]
I am trying to create a FeatureExtraction based on this [description](https://pynwb.readthedocs.io/en/stable/pynwb.ecephys.html#pynwb.ecephys.FeatureExtraction), but I don't understand how to call the electrodes. There's a "get_electrode_group", but not one to get electrodes in the nwb file so I don't know how to reference the electrodes that I created. Additionally, once the FeatureExtraction is created does it automatically get pulled into the nwb file? I saw in some cases you have to reference the nwb file, but when creating a feature extraction there's no reference to the nwb file so how does it get put in the nwb file?

This is what my code for the FeatureExtraction looks like
```
featureExtraction = []
for jdx in range(0,1):        
    featureExtraction.append(FeatureExtraction(
                    electrodes=[jdx],
                    description=['neuralfeature','samplewidth'],
                    times=times,
                    features=[sampleWidths, neuralFeatures]
                    ))
```
I'm appending to a list called featureExtraction becasue I don't understand how to create a new feature as a row in the nwb file. Once I see how to attach the FeatureExtraction to the nwb file this code will be adjusted. Ultimately an example/tutorial with the FeatureExtraction would be helpful. I did not see one anywhere. I looked in [the documentation book](https://buildmedia.readthedocs.org/media/pdf/pynwb/stable/pynwb.pdf) as well as the [website](https://pynwb.readthedocs.io/en/stable/pynwb.ecephys.html#pynwb.ecephys.FeatureExtraction).

This is what my nwb file looks like
<img width="1807" alt="image" src="https://github.com/NeurodataWithoutBorders/helpdesk/assets/37941857/bfca5200-7555-4cd8-93fe-2eed73f3d65c">

**Comment by CodyCBakerPhD** [comment]
Going to ping @oruebel @bendichter @rly on this one

I'm not personally aware of the answer or intention of that neurodata type

**Comment by oruebel** [comment]
> I'm not personally aware of the answer or intention of that neurodata type

`FeatureExtraction` is (to the best of my knowledge) intended for storing features extracted from electrical recordings stored in  `SpikeEventSeries` or `ElectricalSeries`, where features may be from dimension reduction, such as PCA. @rly @bendichter may be more familiar with `FeatureExtraction` than I am, but I'll do my best. Hopefully the below will help answer at least some of the questions.

> There's a "get_electrode_group", but not one to get electrodes in the nwb file s

I believe the electrodes argument of FeatureExtraction behaves the same way as in ElectrodeGroup (the tutorial here shows an example for how this it works with ElectricalSeries https://pynwb.readthedocs.io/en/stable/tutorials/domain/ecephys.html#extracellular-recordings). I think [`NWBFile.create_electrode_table_region`](https://pynwb.readthedocs.io/en/stable/pynwb.file.html#pynwb.file.NWBFile.create_electrode_table_region) is the function to help with creating `DynamicTableRegion` references to the electrodes table. 

> how to create a new feature as a row in the nwb file

Since you are using multiple `FeatureExtraction` objects, I believe the approach to add them to the NWBFile should be to create a [`ProcessingModule`](https://pynwb.readthedocs.io/en/stable/tutorials/general/plot_file.html#processing-modules). Something along the line of:

```
feature_module = nwbfile.create_processing_module(name="feature_extractions", description="features extracted from ......")
feature_module.add(......my FeatureExtraction object....)
```

> Ultimately an example/tutorial with the FeatureExtraction would be helpful.

Agreed

**Reply by ahabdallah27** [reply]
Thanks @oruebel, that was very helpful, but let me take a step back for a moment and explain my inputs and goal as it may not align with the process. I have about 1000 trials of a 2D spiking band power array each for each channel of my 96 channel Utah Array of varying time lengths (t). Therefore, I have a 1000x2x96xt array. How and where do I input that data in the nwb file? I assumed the FeatureExtraction would be the best place for it. Note that I am not inputting any raw neural data into the nwb file.

By referencing the whole electrode table, does that imply I cannot split the SBP by channel? In the unit table, I split the units by electrode so I was hoping (assuming) to do the same here.

Unit table for reference below:
<img width="1818" alt="image" src="https://github.com/NeurodataWithoutBorders/helpdesk/assets/37941857/cf40f624-3372-4a89-807f-f9f11cf9fedd">

Now that I feel less confident in my process, can you guide me into placing the following data into the nwb file?
- Raw Finger Values: raw readings from the finger sensors. This is used to create Finger Angles and has the shape 1000xfxt.
- Finger Angles: finger flexion of each finger (f). Has shape 1000xfxt.
- Position: Positions output at each timestep from a decoder. This has 10 values per step so it has the shape 1000x10xt.
- SPB (discussed above and has the shape 1000x2x96xt

**Reply by CodyCBakerPhD** [reply]
When you say the first dimension is `1000` trials, are the timestamps on those trials ascending and relative to the session start time, or are they aligned to each individual trial start time?

**Reply by ahabdallah27** [reply]
The obs_intervals in the picture of the unit table above are equivalent to the start and end times for the trials I'm referring to. In every trial I have units and the four readings I sent above and they all are in reference to the session start time with those time frames. For example, trial 1 starts at 2ms and ends at 9ms and is sampled at 1ms so it has 8 total elements, trial 2 starts at 12ms and ends at 2883ms and has 2872 total elements. Therefore there isn't a time vector in the data, it looks something like what's below for the Finger Angles. I may have mixed up the dimensions (time should be before fingers). Nonetheless, you can see the outermost dimension is the trial, there isn't an explicit t inside the vector (time is tracked in a standalone list), but these vectors span the length of that time vector, and within each time interval you have the finger positions.

```
[
Trial 1 [        f1 f2 f3 f4
            t=2 [1 2 3 4],
            t=3 [34 3 19 3]
            t=4 [11 25 6 87],
            t=5 [3 3 93 8]
            t=6 [5 6 7 8],
            t=7 [18 3 1 9]
            t=8 [11 12 22 24],
            t=9 [99 12 64 20]
           ],
Trial 2 [        f1   f2     f3  f4
            t=12 [19 324 40 98],
            ...
            t=2883 [17 250 43 18],
           ]
]
```

**Reply by rly** [reply]
> `FeatureExtraction` is (to the best of my knowledge) intended for storing features extracted from electrical recordings stored in `SpikeEventSeries` or `ElectricalSeries`, where features may be from dimension reduction, such as PCA.

The creation of `FeatureExtraction` predates my involvement with NWB, but I agree with what @oruebel wrote, based on what I understand of the documentation and that this data type was defined alongside data types for storing each stage of spike sorting (thresholding = `SpikeEventSeries`, `EventDetection`, and `EventWaveform`; feature extraction / PCA -> `FeatureExtraction`; clustering -> now deprecated `Clustering`).

**Reply by rly** [reply]
@ahabdallah27 It looks like you are working with data that has already been segmented into trials and you do not want to include the raw data that includes data between trials. First, I want to recommend that if you have the raw data that you also include that data. It may not be useful for your analysis, but someone else may want to analyze the data between trials or have continuous spiking or finger sensor data. 

Storing such data in the array formats that you currently have them in is not well-supported in NWB. The generally recommended approach is to store all times relative to the session start time, and store the data in continuous time series objects. In this case, you would store each array in a `TimeSeries` object and use `NWBFile.invalid_times` and `Units.obs_intervals` to define when data exists. 

> Raw Finger Values: raw readings from the finger sensors. This is used to create Finger Angles and has the shape 1000xfxt.

For example, you could store the raw finger values in a `TimeSeries` that is time x finger (or use an `ElectricalSeries` if these data are coming from electrical sensors and the electrodes are defined in `nwbfile.electrodes`). Data would be missing for the time between the trials if you do not have them. The start and end times of each trial would be stored in `nwbfile.trials`, and you can additionally specify in `NWBFile.invalid_times` that no data exists in each of the periods outside of the trials. To reconstruct your Raw Finger Values array, you would iterate through the 1000 rows in `nwbfile.trials`, find the corresponding index of the trial start time in the `TimeSeries` (you can cache these indices in the `timeseries` column of `nwbfile.trials`) and extract data for all fingers and samples from the `TimeSeries` until you get to the trial end time. 

You may also want to consider [creating an NWB extension](https://nwb-overview.readthedocs.io/en/latest/extensions_tutorial/extensions_tutorial_home.html) to create new data types for these data in a formal way. You could then document the axes, restrict the ranges of values and units of measurement, provide additional metadata, etc. Then you and other people in the community could use this extension to store similar data and build analysis and visualization tools that leverage this common format.

**Reply by rly** [reply]
I am just catching up with various discussions after being away from github.

@CodyCBakerPhD 's recommendation in https://github.com/NeurodataWithoutBorders/helpdesk/discussions/71 to store the trial-segmented data in a custom ragged column of the `nwbfile.trials` table is another valid approach (instead of storing them in `TimeSeries` objects with gaps between the actual data).

**Reply by ahabdallah27** [reply]
Which method is preferred? I am only missing 2 or 3 milliseconds between trials so it's not significant. The reason is because we don't actually collect the data for that time, not that we throw the data away to create the trials. Is it okay for us to have a discontinuous timestamps in the TimeSeries? In any case, with the TimeSeries structure, we are saving the same data repeatedly and it's currently about 2GB. Is there any way to reference the same data as a pointer of sorts so that it doesn't have to be saved uniquely?

**Reply by rly** [reply]
If you have most of the data between trials, I would say that storing the data as a `TimeSeries` is better. 

It is OK to have discontinuous timestamps - you would just pass an array of timestamps.

Yes, if you can link the data or timestamps of one TimeSeries to another.
```python
data1 = np.arange(1000).reshape((100, 10))
data2 = np.arange(1000).reshape((100, 10))
timestamps = np.arange(100)
ts1 = TimeSeries(name="test_timeseries1", data=data1, unit="volts", timestamps=timestamps)
ts2 = TimeSeries(name="test_timeseries2", data=data2, unit="volts", timestamps=ts1)  # creates a link to ts1.timestamps so that it is not duplicated in the file
```
The same can be done for the `data` field, but I am not sure why you would want to duplicate your data. Does that answer your question?

**Reply by ahabdallah27** [reply]
When I do this I get an error when I try to read the file back (or view it in my Jupyter Notebook right after running the lines below).

```
COMPRESSION_OPTIONS = 4

RawInputValsTimeSeries = TimeSeries( #preferably SpatialSeries, but needs reference frame
    name="Raw Finger Values",
    data=H5DataIO(data=SliceableDataChunkIterator(data=np.array(RawInputVals)), compression="gzip", compression_opts=COMPRESSION_OPTIONS),
    unit="V",
    timestamps=H5DataIO(data=SliceableDataChunkIterator(data=np.array(ExperimentTime)), compression="gzip", compression_opts=COMPRESSION_OPTIONS),
    conversion=1.0
)

FingerAnglesTIMRLTimeSeries = TimeSeries( #preferably SpatialSeries, but needs reference frame
    name="Finger Angles",
    data=H5DataIO(data=SliceableDataChunkIterator(data=np.array(FingerAnglesTIMRL)), compression="gzip", compression_opts=COMPRESSION_OPTIONS),
    unit="Decimal Percent",
    timestamps=RawInputValsTimeSeries,
    conversion=1.0
)
```

<img width="1186" alt="image" src="https://github.com/NeurodataWithoutBorders/helpdesk/assets/37941857/4f732d6f-040e-43eb-88bf-8233b2634adb">


I thought I may have a problem with my code so I tried your exact lines in my nwbfile and got the same error:
<img width="1188" alt="image" src="https://github.com/NeurodataWithoutBorders/helpdesk/assets/37941857/441606d5-2c27-49ff-9099-1adc94048cbb">

**Reply by rly** [reply]
What version of pynwb and hdmf are you using? We found a bug when printing the nwbfile in jupyter notebooks that looks like what you are encountering. That bug should be resolved in the latest versions of both packages. Please update pynwb and hdmf and let me know if you still encounter the issue.

**Reply by ahabdallah27** [reply]
I was running 2.5.0. Updated to 2.6.0 and it works now, thanks!

Can a combination of the compression and referencing columns be done in the units table too? I get errors when I try either individually. I'm trying to optimize my file size at the moment. The observation intervals for every trial is the same so that's why I want to have a reference for them.

Only compression:
<img width="1643" alt="image" src="https://github.com/NeurodataWithoutBorders/helpdesk/assets/37941857/e0c9b5d8-a1c0-4283-b7bb-e9437f5b0c7b">

Only referencing obs_intervals:
<img width="1538" alt="image" src="https://github.com/NeurodataWithoutBorders/helpdesk/assets/37941857/c3445610-6a57-493c-8471-3e353c9693aa">

**Reply by oruebel** [reply]
`nwbfile.add_unit` adds a single row to the units table. However, the data is represented in the table as columns. That is, for compression you will want to apply `H5DataIO` to the column, rather than when adding individual elements/rows. Similarly, `DataChunkIterator` is something you would use for the whole column, rather than for elements with `add_unit`. 

If you want to use a `DataChunkIterator` (or `H5DataIO`) with the units table, I believe you would not use the `nwbfile.add_unit` function, but instead would set the data for the column as a whole. I'm not sure how, right now, @rly do you have an example how this looks?

Another way to add compression to columns of the units table, I think would be to construct the table as usual via `nwbfile.add_unit` and then calling `set_data_io` on the columns you would want to compress after you are done constructing te table. E.g., `nwbfile.units.obs_intervals.set_data_io(...)` and if you also want to warp the index on that column, then also  `nwbfile.units.obs_intervals_index.set_data_io(...)`. The docs for `set_data_io` is at https://hdmf.readthedocs.io/en/stable/hdmf.container.html#hdmf.container.Data.set_dataio

**Reply by ahabdallah27** [reply]
Thanks!

**Reply by ahabdallah27** [reply]
Is there an easy to to convert the TimeSeries into a DataFrame? I see documentation on converting unit and trials to a dataframe, but not TimeSeries. I have a bunch of TimeSeries saved in acquisition and some may be saved in a processing module. How do I convert all of the TimeSeries in acquisition in a single DataFrame and all of the ones in the processing module in another single DataFrame?

**Reply by ahabdallah27** [reply]
This was my attempt, but it is taking a little less than a second per iteration (through 7 keys) so this method isn't scalable and therefore practical. Furthremore, I noticed I had to read the datafile and was unable to use the below method from a created nwbfile object because of the compression that I applied. Is there a way to convert a compressed TimeSeries into a DataFrame?

```
nwbTimeSeriesDict = dict()
for key in read_nwbfile.acquisition.keys():
    nwbTimeSeriesDict[key] = []
    dataLen = len(read_nwbfile.acquisition[key].data)

for i in range(0, dataLen):
    for key in read_nwbfile.acquisition.keys():
        nwbTimeSeriesDict[key].append(read_nwbfile.acquisition[key].data[i])

pd.DataFrame.from_dict(nwbTimeSeriesDict)
```

**Reply by CodyCBakerPhD** [reply]
First thing first; the `units` and `trials` neurodatatypes are `DynamicTables` and `TimeIntervals` (which descends from `DynamicTable` respectively) and so are more 'table-like' which is why they have an inherent `.to_dataframe() -> pandas.DataFrame` method

One thing to know about a `pandas.DataFrame` is that is that, like a `numpy.array`, it loads all data into memory

So for `TimeSeries`, while you can certainly form a tabular structure out a single 2D one by  of several of them if they have consistent shapes, this is not usually recommended based on how heavy they _can_ be - there are many NWB files on DANDI that if you attempted to load them into memory like this would overload and crash your system (I know from experience 😆)

So with that addendum, let's look at performance...

You should generally avoid iterating single vales of any array-like entity in Python if the equivalent operation can be done with slicing

So instead of doing

```python
for i in range(0, dataLen):
    for key in read_nwbfile.acquisition.keys():
        nwbTimeSeriesDict[key].append(read_nwbfile.acquisition[key].data[i])
```

you should just be able to do

```python
for key in read_nwbfile.acquisition.keys():
    nwbTimeSeriesDict[key].append(read_nwbfile.acquisition[key].data[:])
```

Of course, pandas has compatibility with `h5py.Dataset` objects anyway so you shouldn't even need to slice them at all. My best attempt to replicate this, without seeing your data, is

```python
import pandas
from pynwb.testing.mock.base import mock_TimeSeries

list_of_timeseries = [mock_TimeSeries(), mock_TimeSeries(), mock_TimeSeries()]
pandas.DataFrame.from_dict({time_series.name: time_series.data for time_seriesin list_of_timeseries})
```

**Reply by ahabdallah27** [reply]
Thanks, it worked! I needed to also convert my np array into a list before turning it inserting them into a dictionary.
